Kitodo/src/main/java/org/kitodo/production/services/index/IndexingService.java

```diff
@@ -286,50 +278,114 @@ private int getNumberOfDatabaseObjects(ObjectType objectType) throws DAOExceptio
-                        pollingChannel.send(INDEXING_STARTED_MESSAGE + currentIndexState);
-                    }
-                    indexerThread = new Thread(worker);
-                    indexerThread.setName("Indexing " + worker.getIndexedObjects() + " of type " + type);
-                    indexerThread.setDaemon(true);
-                    indexerThread.start();
-                    indexerThread.join();
-                    break;
-                } else {
-                    logger.debug("Cannot start '{}' indexing while a different indexing process running: '{}'", type,
-                            this.currentIndexState);
-                    Thread.sleep(PAUSE);
-                    attempts++;
+    /**
+     * Wait and check whether index worker threads have finished or failed. 
+     * 
+     * <p>Also trigger UI updates every second.</p>
+     * 
+     * <p>This method is executed in the `IndexAllThread`.</p>
+     * 
+     * @param type the object type currently being indexed
+     * @param futures a list of futures allowing to check the status of the worker threads
+     * @param pollingChannel the UI polling channel
+     */
+    private void waitWhileIndexing(ObjectType type, List<Future<?>> futures, PushContext pollingChannel) {
+        while (true) {
+            // check whether all jobs are done
+            boolean done = true;
+            boolean failed = false;
+            for (Future<?> future : futures) {
+                if (!future.isDone()) {
+                    done = false;
                 }
+                if (future.isCancelled()) {
+                    failed = true;
+                }
+            }
+
+            if (done) {
+                // indexing has completed
+                logger.info("indexing of " + type.toString() + " finished successfully");
+                currentIndexState = ObjectType.NONE;
+                currentState = IndexStates.INDEXING_SUCCESSFUL;
+                objectIndexingStates.put(type, IndexStates.INDEXING_SUCCESSFUL);
+                break;
+            }
+
+            if (failed) {
+                logger.info("indexing of " + type.toString() + " failed, cleaning up");
+                currentIndexState = ObjectType.NONE;
+                currentState = IndexStates.INDEXING_FAILED;
+                objectIndexingStates.put(type, IndexStates.INDEXING_FAILED);
+
+                for (Future<?> future : futures) {
+                    future.cancel(true);
+                }
+                break;
+            }
+
+            // send update
+            pollingChannel.send(INDEXING_STARTED_MESSAGE + type);
+
+            // wait a bit
+            try {
+                Thread.sleep(PAUSE);
             } catch (InterruptedException e) {
-                Helper.setErrorMessage(e.getLocalizedMessage(), logger, e);
-                Thread.currentThread().interrupt();
+                // ignore
```

Comment:
Same applies here like in the above construct for ignoring the `InterruptedException`.